contributors/devel/sig-instrumentation/trace-instrumentation.md

```diff
@@ -0,0 +1,75 @@
+Kubernetes components falls into one of two categories:
+
+1. Spans for incoming or outgoing network calls
+2. Spans when initiating new work, such as reconciling an object, which may result in network calls.
+
+For network-based telemetry, Kubernetes components should use OpenTelemetry
+instrumentation libraries for
+[HTTP](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp) and
+[gRPC](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc).
+
+**Note:** When creating spans at the start of reconciling an Object, only
+create the span changes are actually required. Avoid creating "empty" spans
+which simply compare the desired and actual state of an object without
+performing any real work, or making any network requests.
+
+### Configuration and Setup
+
+Kubernetes components should expose a flag, `--tracing-config-file`, which accepts a
+[TracingConfiguration](https://kubernetes.io/docs/reference/config-api/apiserver-config.v1beta1/#apiserver-k8s-io-v1beta1-TracingConfiguration)
+object. The `component-base/tracing` library provides a `NewProvider()` helper
+to convert a TracingConfiguration to a TracerProvider, which can be used to
+record spans. Components should avoid using OpenTelemetry globals, and instead
+pass the configured TracerProvider to libraries where they are used. Components
+should use the W3C Traceparent and Baggage propagators, as provided by the
+`Propagators()` helper.
+
+### Context Propagation
+
+Generally, components should not interact directly with OpenTelemetry
+Propagators, other than by passing them to libraries. Context propagation
+across network boundaries is handled by the
+[HTTP](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp) and
+[gRPC](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc)
+network client and server instrumentation libraries.
+
+Components need to propagate Golang's `context.Context` from incoming network
+calls or spans from the initiation of new work to any outgoing network calls to
+ensure spans are properly connected into traces.
+
+### Naming and Style
+
+Follow the OpenTelemetry [guidelines for span naming](https://opentelemetry.io/docs/specs/otel/trace/api/#span), and the OpenTelemetry [guidelines for attributes](https://opentelemetry.io/docs/specs/semconv/general/attribute-naming/).
+
+### Tracing stability
+
+Tracing instrumentation in Kubernetes components does not currently have
+stability guarantees, but component owners should be aware of which changes are
+breaking to users so such changes are done with proper consideration. In
+particular, is is breaking for users for a component to stop propagating
+context in a way that breaks parent/child relationships for spans, to remove
+spans without replacement, or to remove an attribute from a span without
+replacement. Component owners should not treat general modification spans
```

Comment:
Is this correct that renaming the span is not breaking change for user?

Reply:
It is breaking for users in the same way that changing a log message or attributes is breaking for users.  I'm arguing that while tracing is new in kubernetes, we should not treat span names or attributes as a stable API, and should be clear to users that (like log messages), span names, attributes or structure shouldn't be relied on. In the future, we can always change this to make more stability requirements if there is a need.